beater/v2_handler.go

```diff
@@ -0,0 +1,297 @@
+	var err error
+	var rawModel map[string]interface{}
+
+	eventables := []transform.Transformable{}
+	for i := 0; i < batchSize && err == nil; i++ {
+		rawModel, err = reader.Read()
+		if err != nil && err != io.EOF {
+			switch e := err.(type) {
+			case decoder.ReadError:
+				response.AddWithMessage(ServerError, 1, e.Error())
+				// return early, we can't recover from a read error
+				return eventables, true
+			case decoder.JSONDecodeError:
+				response.AddWithOffendingDocument(InvalidJSONErr, e.Error(), reader.LastLine())
+				response.Invalid++
+			}
+		}
+
+		if rawModel != nil {
+			tr, err := v.handleRawModel(rawModel)
+			if err != nil {
+				response.AddWithOffendingDocument(SchemaValidationErr, err.Error(), reader.LastLine())
+				response.Invalid++
+			}
+			eventables = append(eventables, tr)
+		}
+	}
+
+	return eventables, reader.IsEOF()
+}
+func (v *v2Handler) readMetadata(r *http.Request, ndReader *decoder.NDJSONStreamReader) (*metadata.Metadata, error) {
+	// first item is the metadata object
+	rawModel, err := ndReader.Read()
+	if err != nil {
+		return nil, err
+	}
+
+	rawMetadata, ok := rawModel["metadata"].(map[string]interface{})
+	if !ok {
+		return nil, errUnrecognizedObject
+	}
+	// augment the metadata object with information from the request, like user-agent or remote address
+	reqMeta, err := v.requestDecoder(r)
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range reqMeta {
+		utility.InsertInMap(rawMetadata, k, v.(map[string]interface{}))
+	}
+
+	// validate the metadata object against our jsonschema
+	err = validation.Validate(rawMetadata, metadata.ModelSchema())
+	if err != nil {
+		return nil, err
+	}
+
+	// create a metadata struct
+	metadata, err := metadata.DecodeMetadata(rawMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return metadata, nil
+}
+
+func (v *v2Handler) handleRequestBody(r *http.Request, ndReader *decoder.NDJSONStreamReader, report reporter) *StreamResponse {
+	requestTime := requestTime(r)
+	resp := &StreamResponse{}
+
+	metadata, err := v.readMetadata(r, ndReader)
+
+	// no point in continuing if we couldn't read the metadata
+	if err != nil {
+		switch e := err.(type) {
+		case decoder.ReadError:
+			resp.AddWithMessage(ServerError, 1, e.Error())
+		case decoder.JSONDecodeError:
+			resp.AddWithOffendingDocument(InvalidJSONErr, err.Error(), ndReader.LastLine())
+		default:
+			resp.AddWithOffendingDocument(SchemaValidationErr, err.Error(), ndReader.LastLine())
+		}
+
+		return resp
+	}
+
+	tctx := &transform.Context{
+		RequestTime: requestTime,
+		Config:      v.tconfig,
+		Metadata:    *metadata,
+	}
+
+	for {
+		transformables, done := v.readBatch(batchSize, ndReader, resp)
+		if transformables != nil {
+			err := report(r.Context(), pendingReq{
+				transformables: transformables,
+				tcontext:       tctx,
+			})
+
+			if err != nil {
+				if strings.Contains(err.Error(), "publisher is being stopped") {
+					resp.Add(ShuttingDownErr, 1)
+					return resp
+				}
```

Comment:
Did you miss to cover the [`ctx.Error()`](https://github.com/elastic/apm-server/blob/master/beater/pub.go#L124)?

Reply:
It turns out that this is actually equivalent to what we have today. 

Unfortunately, we throw away `ctx.Error()` in v1:  https://github.com/elastic/apm-server/blob/master/beater/handlers.go#L156-L161.

I'll check if we can safely change it in both places so it becomes something similar to `errors.Wrap(ctx.Error(), "queue is full")`